src/core/map/tile.rs

```diff
@@ -166,3 +166,127 @@ impl<IO: EngineIO> TileTypes<IO> {
 	// 	Ok(idx as TileIdx)
 	// }
 }
+
+#[cfg(test)]
+mod tile_tests {
+	use super::*;
+	use proptest::prelude::*;
+	use std::{convert::Infallible, hash::Hasher, path::PathBuf};
+
+	#[derive(Debug, Default, Eq, PartialEq)]
+	struct DummyIO {}
+
+	impl EngineIO for DummyIO {
+		type ReadError = Infallible;
+		type Read = &'static [u8];
+
+		fn read(&mut self, _: PathBuf) -> Result<Self::Read, Self::ReadError> {
+			Ok(b"")
+		}
+
+		type TileInterface = ();
+
+		fn blank_tile_interface() -> Self::TileInterface {}
+
+		type TileAddedError = Infallible;
+
+		fn tile_added(
+			&mut self,
+			_: usize,
+			_: &mut TileType<Self>,
+		) -> Result<(), Self::TileAddedError> {
+			Ok(())
+		}
+	}
+
+	impl PartialEq for TileType<DummyIO> {
+		fn eq(&self, other: &Self) -> bool {
+			self.name == other.name
+		}
+	}
+	impl Eq for TileType<DummyIO> {}
+
+	impl std::hash::Hash for TileType<DummyIO> {
+		fn hash<H: Hasher>(&self, state: &mut H) {
+			self.name.hash(state);
+		}
+	}
+
+	fn tiletype_strategy_generator(regex: &str) -> BoxedStrategy<TileType<DummyIO>> {
+		prop::string::string_regex(regex)
+			.expect("failed to generate strategy from regex")
+			.prop_map(|s| TileType {
+				name: s,
+				interface: (),
+			})
+			.boxed()
+	}
+
+	fn rand_dummy_tiletype_strategy() -> BoxedStrategy<TileType<DummyIO>> {
+		tiletype_strategy_generator(".*")
+	}
+
+	fn non_empty_tiletype_strategy() -> BoxedStrategy<TileType<DummyIO>> {
+		tiletype_strategy_generator(".+")
+	}
+
+	proptest!(
+		#[test]
+		fn first_tiletype_should_be_unique(tt in rand_dummy_tiletype_strategy()) {
+			let mut dummy_io = DummyIO::default();
+			let mut tts = TileTypes::new();
+			if let Err(TileTypesError::DuplicateTileTypeName(s)) = tts.add_tile(&mut dummy_io, tt) {
+				prop_assert!(false, "TileType {} marked as duplicate, but it is the only one added", s);
+			}
+		}
+	);
+
+	proptest!(
+		#[test]
+		fn non_empty_tiletypes_are_valid(tt in non_empty_tiletype_strategy()) {
+			let mut dummy_io = DummyIO::default();
+			let mut tts = TileTypes::new();
+			let name = tt.name.clone();     // I really don't want to do this, but add_tile consumes
+			if let Err(TileTypesError::InvalidTileTypeData(s)) = tts.add_tile(&mut dummy_io, tt) {
+				prop_assert!(false, "TileType {} marked invalid because {}", name, s);
+			}
+		}
+	);
+
+	proptest!(
+		#![proptest_config(ProptestConfig::with_cases(30))]
+		#[test]
+		fn many_valid_tiletypes_get_accepted(
+			// I would use 2..TileIdx::MAX for the size but it pins the cpu
+			tt_set in prop::collection::hash_set(non_empty_tiletype_strategy(), 2..500)
```

Comment:
What exactly does the `2..500` do in this?  Is that the possible size that it fills the set with?  If so then 65k (`MAX`) would indeed be quite a lot and I could definitely see it being slow...

Reply:
yeah it's basically "make a hashset of length between 2 and 500 where each element is generated from this strategy". I originally had it set to `TileIdx::Max` but it would take a really long time